### DISCUSSION TITLE
always add new role above the child group

### DISCUSSION THREAD

**Discussion by sherifawad** [react-querybuilder/react-querybuilder#766] [Q&A] [answered]
in main group if there is a nested group and added a rule in the main it become under the nested group is there a way to stick the rules together without group separation
![Screenshot 2024-08-22 222830](https://github.com/user-attachments/assets/abcf4923-f8db-4fef-a2a9-9b57904f641b)

**Comment by jakeboone02** [comment]
To restate the question more generally, would you want to keep all rules at the top and all subgroups at the bottom? Like, "+ Group" would add a group to the very bottom but "+ Rule" would insert after any top-level rules but before any subgroups?

**Reply by sherifawad** [reply]
Thank you for clarifying the question regarding the organization of rules and subgroups.
I would prefer to keep all top-level rules at the top and have all subgroups listed at the bottom. This structure allows for easier navigation and understanding of the main rules before diving into the specific subgroups.

**Reply by jakeboone02** [reply, marked as answer]
Working example:  https://codesandbox.io/p/devbox/r5tm73?file=%2Fsrc%2FApp.tsx

```tsx
const AddRule = (props: ActionWithRulesAndAddersProps) => {
  const groupIndex = (props.ruleOrGroup as RuleGroupType).rules.findIndex(r =>
    isRuleGroup(r)
  );

  if (groupIndex === -1) {
    return <ActionElement {...props} />;
  }

  const addRule = () => {
    props.schema.dispatchQuery(
      insert(props.schema.getQuery()!, defaultRule, [...props.path, groupIndex])
    );
  };

  return <ActionElement {...props} handleOnClick={addRule} />;
};

export const App = () => {
  const [query, setQuery] = useState(initialQuery);

  return (
    <QueryBuilder
      fields={fields}
      query={query}
      onQueryChange={setQuery}
      controlElements={{ addRuleAction: AddRule }}
    />
  );
};

```

**Reply by jakeboone02** [reply]
@sherifawad if that worked for you, can you mark it as the answer?
